File: com/DataStructure/Python/BinarySearchTree.py
from collections import deque


class TreeNode:

    def __init__(self, val=None):
        self.val = val
        self.left = None
        self.right = None


class BinarySearchTree:

    def __init__(self):
        self.root = None

    def getRoot(self):
        return self.root

    def addNode(self, val):
        newNode = TreeNode(val)
        fNode = self.root
        while True:
            if not self.root:
                self.root = newNode
                break
            elif val <= fNode.val:
                if not fNode.left:
                    fNode.left = newNode
                    break
                else:
                    fNode = fNode.left
            else:
                if not fNode.right:
                    fNode.right = newNode
                    break
                else:
                    fNode = fNode.right

    def printTree(self, fNode):
        if fNode:
            print(fNode.val, end=" ")
            self.printTree(fNode.left)
            self.printTree(fNode.right)

    def printTreeBfs(self, fNode):
        q = deque()
        q.append(fNode)
        while q:
            current = q.popleft()
            if current:
                print(current.val, end=" ")
                q.append(current.left)
                q.append(current.right)
        print()


if __name__ == "__main__":
    newBinaryTree = BinarySearchTree()
    newBinaryTree.addNode(10)
    newBinaryTree.addNode(15)
    newBinaryTree.addNode(19)
    newBinaryTree.addNode(17)
    newBinaryTree.addNode(11)
    newBinaryTree.printTreeBfs(newBinaryTree.root)

File: com/DataStructure/BinarySearchTree.java
package com.DataStructure;

import java.util.Stack;
import java.util.LinkedList;
import java.util.Queue;

public class BinarySearchTree {

	TreeNode root;

	public void addNode(int val) {
		TreeNode newNode = new TreeNode(val);
		TreeNode fNode = root;

		while (true) {
			if (root == null) {
				root = newNode;
				break;
			} else if (val <= fNode.val) {
				if (fNode.left == null) {
					fNode.left = newNode;
					break;
				} else {
					fNode = fNode.left;
				}
			} else {
				if (fNode.right == null) {
					fNode.right = newNode;
					break;
				} else {
					fNode = fNode.right;
				}
			}
		}
	}

	public TreeNode getRoot() {
		return root;
	}

	public TreeNode search(int val) {
		TreeNode fNode = root;
		while (fNode != null && fNode.val != val) {
			if (val < fNode.val) {
				fNode = fNode.left;
			} else {
				fNode = fNode.right;
			}
		}

		if (fNode == null)
			return null;

		return fNode;
	}

	public void printTreeR(TreeNode fNode) {
		if (fNode != null) {
			System.out.print(fNode.val + " ");
			printTreeR(fNode.left);
			printTreeR(fNode.right);
		}
	}

	public void printTreeI(TreeNode fNode) {
		if (fNode == null)
			return;
		Stack<TreeNode> s = new Stack<TreeNode>();
		s.push(fNode);
		while (!s.isEmpty()) {
			fNode = s.pop();
			System.out.print(fNode.val + " ");
			if (fNode.right != null) {
				s.push(fNode.right);
			}
			if (fNode.left != null) {
				s.push(fNode.left);
			}
		}
	}

	public void printTreeLevelOrder(TreeNode fNode) {
		if (fNode == null) {
			System.out.println("Empty Tree!");
			return;
		}
		Queue<TreeNode> queue = new LinkedList<>();
		queue.add(fNode);
		while (!queue.isEmpty()) {
			TreeNode current = queue.poll();
			System.out.print(current.val+" ");
			if (current.left != null) {
				queue.add(current.left);
			}
			if (current.right != null) {
				queue.add(current.right);
			}
		}
		System.out.println();
	}

	public static void main(String args[]) {
		System.out.println("Hello World!");
		BinarySearchTree newtree = new BinarySearchTree();
		newtree.addNode(3);
		newtree.addNode(20);
		newtree.addNode(9);
		newtree.addNode(15);
		newtree.addNode(7);
		newtree.printTreeLevelOrder(newtree.getRoot());
	}

	public class TreeNode {
		public int val;
		public TreeNode right;
		public TreeNode left;

		public TreeNode(int val) {
			this.val = val;
		}
	}
}